scripts/schemagen.go

```diff
@@ -0,0 +1,277 @@
+		// e.g. compute, instance -> ComputeInstance
+		TypeName: strings.ToUpper((*api)[0:1]) + (*api)[1:] + *resource,
+		Fields:   fields,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = exec.Command("gofmt", "-w", fileName).Run()
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func generateFields(schemas map[string]discovery.JsonSchema, property string) map[string]string {
+	fields := make(map[string]string, 0)
+
+	for k, v := range schemas[property].Properties {
+		content, err := generateField(schemas, k, v, false)
+		if err != nil {
+			log.Printf("ERROR: %s", err)
+		} else {
+			fields[underscore(k)] = content
+		}
+	}
+
+	return fields
+}
+
+func generateField(schemas map[string]discovery.JsonSchema, field string, v discovery.JsonSchema, isNested bool) (string, error) {
+	s := &schema.Schema{
+		Description: v.Description,
+	}
+	if field != "" {
+		setProperties(v, s)
+	}
+
+	// JSON field types: https://tools.ietf.org/html/draft-zyp-json-schema-03#section-5.1
+	switch v.Type {
+	case "integer":
+		s.Type = schema.TypeInt
+	case "number":
+		s.Type = schema.TypeFloat
+	case "string":
+		s.Type = schema.TypeString
+	case "boolean":
+		s.Type = schema.TypeBool
+	case "array":
+		s.Type = schema.TypeList
+		elem, err := generateField(schemas, "", *v.Items, true)
+		if err != nil {
+			return "", fmt.Errorf("Unable to generate Elem for %q: %s", field, err)
+		}
+		s.Elem = elem
+	case "object":
+		s.Type = schema.TypeMap
+	case "":
+		s.Type = schema.TypeList
+		s.MaxItems = 1
+
+		elem := "&schema.Resource{\nSchema: map[string]*schema.Schema{\n"
+		m := generateFields(schemas, v.Ref)
+		fieldNames := []string{}
+		for k, _ := range m {
+			fieldNames = append(fieldNames, k)
+		}
+		sort.Strings(fieldNames)
+		for _, k := range fieldNames {
+			elem += fmt.Sprintf("%q: %s,\n", k, m[k])
+		}
+		elem += "},\n}"
+
+		if isNested {
+			return elem, nil
+		}
+		s.Elem = elem
+	default:
+		return "", fmt.Errorf("Unable to process: %s %s", field, v.Type)
+	}
+
+	return schemaCode(s, isNested)
+}
+
+func setProperties(v discovery.JsonSchema, s *schema.Schema) {
+	if v.ReadOnly || strings.HasPrefix(v.Description, "Output-only") {
+		s.Computed = true
+	} else {
+		if v.Required || strings.HasPrefix(v.Description, "Required") {
+			s.Required = true
+		} else {
+			s.Optional = true
+		}
+	}
+
+	s.ForceNew = true
+}
+
+func schemaCode(s *schema.Schema, isNested bool) (string, error) {
+	buf := bytes.NewBuffer([]byte{})
+	err := schemaTemplate.Execute(buf, struct {
+		Schema   *schema.Schema
+		IsNested bool
+	}{
+		Schema:   s,
+		IsNested: isNested,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// shamelessly borrowed from https://github.com/radeksimko/terraform-gen/blob/master/internal/util/util.go
+func underscore(name string) string {
+	var words []string
+	var camelCase = regexp.MustCompile("(^[^A-Z]*|[A-Z]*)([A-Z][^A-Z]+|$)")
```

Comment:
I found a different regexp that's a bit more robust so I replaced this with that. The `externalIPs` case is a pretty extreme corner-case because of the plural, so I just hardcoded in handling of plurals.